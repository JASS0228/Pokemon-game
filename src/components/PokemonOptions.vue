<template>
	<div class="options-container">
		<ul v-if="!selected">
			<li
				v-for="{ name, id } in pokemons"
				:key="id"
				@click="$emit('selectionPokemon', id)">
				{{ name }}
			</li>
		</ul>
	</div>
</template>
<script>
export default {
	props: {
		pokemons: {
			type: Array,
			required: true,
		},
		selected: {
			type: Boolean,
		},
	},
	methods: {},
}
</script>

<style scoped>
ul {
	list-style-type: none;
}
li {
	background-color: white;
	border-radius: 5px;
	border: 1px solid rgba(0, 0, 0, 0.2);
	cursor: pointer;
	margin-bottom: 10px;
	width: 250px;
}

li:hover {
	background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
	display: flex;
	justify-content: center;
}
</style>
