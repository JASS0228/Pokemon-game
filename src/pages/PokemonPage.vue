<template>
	<h1 v-if="!pokemon">Espere por favor...</h1>

	<div v-else>
		<h1>Quien es este Pokemon</h1>

		<PokemonPicture
			:pokemonId="pokemon.id"
			:showPokemon="showPokemon" />

		<PokemonOptions
			:pokemons="pokemonArr"
			:selected="showAnswer"
			@selection-pokemon="checkAnswer" />

		<template v-if="showAnswer">
			<h2 class="fade-in">
				{{ message }}
			</h2>
			<button @click="newGame">Nuevo Juego</button>
		</template>
	</div>

	<!-- <PokemonAnswers
		:correct="correctAnswer"
		:incorrect="incorrectAnswer" />

	<PokemonScore :score="score" /> -->
</template>
<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import PokemonAnswers from '@/components/PokemonAnswers.vue'
import PokemonScore from '@/components/PokemonScore.vue'

import getPokemonsOptions from '@/helpers/getPokemonOptions.js'

export default {
	components: {
		PokemonPicture,
		PokemonOptions,
		PokemonAnswers,
		PokemonScore,
	},
	data() {
		return {
			pokemonArr: [],
			pokemon: null,
			showPokemon: false,
			showAnswer: false,
			message: '',
			// correctAnswer: 0,
			// incorrectAnswer: 0,
			// score: 0,
		}
	},
	methods: {
		async mixPokemonArray() {
			this.pokemonArr = await getPokemonsOptions()

			const rndInt = Math.floor(Math.random() * 4)

			this.pokemon = this.pokemonArr[rndInt]
		},
		checkAnswer(selectedId) {
			this.showPokemon = true
			this.showAnswer = true

			if (selectedId === this.pokemon.id) {
				this.message = `Correcto, ${this.pokemon.name}`
				// this.correctAnswer++
				// this.score += 10
			} else {
				this.message = `Fallaste es ${this.pokemon.name}`
				// this.incorrectAnswer++
				// this.score -= 5
			}
		},
		newGame() {
			this.pokemonArr = []
			this.showAnswer = false
			this.showPokemon = false
			this.pokemon = null

			this.mixPokemonArray()
		},
	},
	mounted() {
		this.mixPokemonArray()
	},
}
</script>
