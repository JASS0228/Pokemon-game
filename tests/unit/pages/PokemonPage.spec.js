import { mount } from '@vue/test-utils'
import { pokemons } from '../mocks/pokemons.mock'
import PokemonPage from '@/pages/PokemonPage.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import PokemonPicture from '@/components/PokemonPicture.vue'

describe('PokemonPage component', () => {
	let wrapper

	beforeEach(() => {
		wrapper = mount(PokemonPage)
	})

	test('debe de hacer match con el snapshot', () => {
		expect(wrapper.html()).toMatchSnapshot()
	})

	//Test a mounted de PokemonPage
	test('debe de llamar el mixPokemonArr al montar', () => {
		const mixPokemonArraySpy = jest.spyOn(
			PokemonPage.methods,
			'mixPokemonArray'
		)

		const wrapper = mount(PokemonPage)

		expect(mixPokemonArraySpy).toHaveBeenCalled()
	})

	test('debe de hacer match con la snapshot cuando carga los pokemons', () => {
		const wrapper = mount(PokemonPage, {
			data() {
				return {
					pokemonArr: pokemons,
					pokemon: pokemons[1],
					showPokemon: false,
					showAnswer: false,
					message: '',
				}
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
		const wrapper = mount(PokemonPage, {
			data() {
				return {
					pokemonArr: pokemons,
					pokemon: pokemons[1],
					showPokemon: false,
					showAnswer: false,
					message: '',
				}
			},
		})

		const wrapperPicture = mount(PokemonPicture, {
			props: {
				pokemonId: 2,
				showPokemon: false,
			},
		})

		const wrapperOptions = mount(PokemonOptions, {
			props: {
				pokemons,
			},
		})

		//PokemonPicture debe de existir
		expect(wrapper.find('.pokemon-container').exists()).toBeTruthy()

		//PokemonOptions debe de existir
		expect(wrapper.find('.options-container').exists()).toBeTruthy()

		//PokemonPicture debe tener el atributo pokemonId === 2
		expect(wrapperPicture.componentVM.pokemonId).toBe(2)

		//PokemonOptions debe de tener atributo pokemons to be True
		expect(wrapperOptions.componentVM.pokemons).toStrictEqual(pokemons)
	})

	test('pruebas en checkAnswer', async () => {
		const wrapper = mount(PokemonPage, {
			data() {
				return {
					pokemonArr: pokemons,
					pokemon: pokemons[1],
					showPokemon: false,
					showAnswer: false,
					message: '',
				}
			},
		})

		await wrapper.componentVM.checkAnswer({ selectedId: pokemons[1].id })

		expect(wrapper.find('h2').exists()).toBeTruthy()
	})
})
