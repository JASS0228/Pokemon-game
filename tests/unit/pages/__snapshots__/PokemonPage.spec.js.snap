// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hacer match con el snapshot 1`] = `
<h1>Espere por favor...</h1><!-- <PokemonAnswers
		:correct="correctAnswer"
		:incorrect="incorrectAnswer" />

	<PokemonScore :score="score" /> -->
`;

exports[`PokemonPage component debe de hacer match con la snapshot cuando carga los pokemons 1`] = `
<div>
  <h1>Quien es este Pokemon</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/2.svg" class="hidden-pokemon" alt="Pokemon"></div>
  <div class="options-container">
    <ul>
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div><!--v-if-->
</div><!-- <PokemonAnswers
		:correct="correctAnswer"
		:incorrect="incorrectAnswer" />

	<PokemonScore :score="score" /> -->
`;
